### PR TITLE
feat: Orgs: Auto-redirect after deleting team members

### LINKED ISSUE #15259: Orgs: Auto-redirect after deleting team members
I think it would be cool that if you remove people from your organisation, you can select another team member that replaces that person.

Often links are somewhere inside emails, or deeply added into applications and when offboarding that person it would be nice to select a “replacement”.

I.e.: [acme.cal.com/brian](http://acme.cal.com/brian) stops working there, the admin who is removing brian can select acme.cal.com/anna, so all future links redirect to anna

### PR DESCRIPTION
**Pull request by Amit91848**
## What does this PR do?

- Fixes #15259
- Fixes CAL-3831

https://github.com/calcom/cal.com/assets/74371312/1c1a3245-8e83-415b-9b3d-949d93a0e05c

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

### CODE REVIEW

**Review by sean-brydon — CHANGES_REQUESTED** [review]
Great job at this! I think we can leverage the db schema we already have and take advantage of the "tempOrgRedirect" table. This would eliminate all the code needed to handle the redirect as that is already handled on the booking page already

I'm happy from a UX side this is pretty solid.

**Comment by Amit91848** [comment]
Do you think we should also show whatever redirects have been created for an org, similar to the table in ooo?

**Comment by sean-brydon** [comment]
> Do you think we should also show whatever redirects have been created for an org, similar to the table in ooo?

Thats not a bad idea! Deffo not needed for this feature but would be a nice to have if you have time

**Review by sean-brydon — COMMENTED** [review]
Looks pretty good to me! @hariombalhara may have some more comments as he knows this redirect stuff better than me. 

Would you be down to expand our tests to ensure this is tested correctly? 

`getTemporaryOgRedirect.test.ts` could be expanded here

**Comment by sean-brydon** [comment]
LGTM - tested and it works perfectlly. Thanks for the tests , great job as always @Amit91848

File: packages/features/ee/teams/lib/removeMember.ts [hunk 1]
```diff
@@ -10,10 +12,12 @@ const removeMember = async ({
   memberId,
   teamId,
   isOrg,
+  redirectTo,
 }: {
   memberId: number;
   teamId: number;
   isOrg: boolean;
+  redirectTo?: number;
```
Comment (hariombalhara) [inline packages/features/ee/teams/lib/removeMember.ts:20]: `redirectToUserId` is more clear as `redirectTo` could be `username`, team slug etc.

**Comment by hariombalhara** [comment]
If acme.cal.com/member1 is replaced by acme.cal.com/member2 but acme.cal.com/member2/my-event doesn't exist and /member1/my-event exists, going to /member1/my-event would take to 404 /member2/my-event.

I think we should redirect such 404 pages to the member2 profile page. That would be a special handling in addition to TempOrgRedirect.

To keep things simple we could also **redirect all event-types of member1** to **member2 profile page** from where the user can choose which event suits them. It can't be done through TempOrgRedirect currently. We might have to introduce a new RedirectType which could be `RedirectType.UserToProfile`

Another problem could be that the username could be allocated to some new member later on which could be a problem. So, I think we shouldn't show the user as deleted in this case. They should be shown as 'Replaced by Member2'. Admin can then delete the redirect(e.g. when a new member with same username joins) or replace that with another Member(say Member3).

**Comment by hariombalhara** [comment]
> Not sure wdym by this, where do you want to show Replaced by Member2?

I meant to show it somewhere in Members listing only. Right now it just stops existing. But we can have a placeholder in there, still for this purpose. This is just my thought though.


> I suggested earlier we can show this in a table format where owners and admins can delete and update redirects instead of making it fixed to a single user. Lmk what you think.

Yeah I think something like this would be necessary in the end. As I shared earlier, we can postpone the 'showing in UI' part but we have to keep the data in a table about the replacement, so that UI can be built later over it.


> This pr relies on tempOrgRedirect completely, I also noticed this issue https://github.com/calcom/cal.com/issues/15575 aiming to deprecate tempOrgRedirect. Should I wait till final decisions are made for it before continuing with this pr?
I don't mind introducing a new RedirectType that I shared earlier, as it would make sense in long term as well. So, we can implement it in current table only.


TempOrgRedirect would have all it's data in future too but in a better schema that understands relationship with different entities better. It would be like holding raw redirects due to any reasons. Any redirect on the booking page, it would be aware of it.

**Comment by Amit91848** [comment]
@hariombalhara made changes so that redirect created for member1 to member2 would redirect `/acme.cal/member1` to `/acme.cal/member2` and also redirect `/acme.cal/member1/{event-type}` to `/acme.cal/member2/`. Added a new RedirectType and used that for both. 

Only remaining part is ui. Not sure where to show it tho. Org members table is has infinite scrolling, so it would be a issue to show it below the members table for big orgs.

**Comment by hariombalhara** [comment]
@Amit91848 I will review and test it out soon. It is a tricky thing and it needs to be implemented correctly.


cc @PeerRich for review as well.

**Comment by PeerRich** [comment]
i wonder if we can define a generic "backup" system where if people go OOO or their account gets deleted it will redirect

**Comment by Amit91848** [comment]
Hey @hariombalhara any updates on this?

**Comment by hariombalhara** [comment]
@Amit91848 I would like to keep it on the back burner for now as it isn't a priority and this isn't a trivial solution.

**Comment by anikdhabal** [comment]
Closing it for now as this is not too prio for now

**Pull request closed without merging**